File: source/downloads/code/LeetCode/Python/Longest-Absolute-File-Path.py
class Solution(object):
    def lengthLongestPath(self, input):
        """
        :type input: str
        :rtype: int
        """
        
        hash_map = {0:0} # <目录深度， 当前字符串长度>
        maxlen = 0

        # splitlines以'\n'划分
        for line in input.splitlines():
            # 有几个\t 就是第几层目录
            depth = line.count('\t')
            name = line.lstrip('\t')
            # 找到文件
            if '.' in name:
                maxlen = max(maxlen, hash_map[depth]+len(name))
            else:
                # "1": "/"
                hash_map[depth+1] = hash_map[depth]+len(name)+1
        return maxlen

File: source/downloads/code/LeetCode/Java/Longest-Absolute-File-Path.java
public class Solution {
    public int lengthLongestPath(String input) {
        // 把输入按照\n分开
        String[] parts = input.split("\n");
        
        int[] stack = new int[parts.length+1];
        //stack[i]: 坐标为i的"\t"对应的路径长度
        int maxLen = 0;
        for(String s : parts){
            // lev: 当前s到最后一个\t的长度
            int lev = s.lastIndexOf("\t")+1;
            int curLen = stack[lev+1] = stack[lev]+s.length()-lev+1;
            if(s.contains(".")) maxLen = Math.max(maxLen, curLen-1);
        }
        return maxLen;
    }
}
